qa/tasks/cephfs/test_exports.py

```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+        self.mount_a.create_n_files("1/file", 100, sync=True)
+        subtrees = self._wait_distributed_subtrees(3 * 2, status=self.status, rank="all")
+        for s in subtrees:
+            path = s['dir']['path']
+            if path == '/1':
+                self.assertTrue(s['distributed_ephemeral_pin'])
+                self.assertEqual(s['bal_rank_mask'], "2")
+
+    def test_ephemeral_random_overried_rank_mask(self):
+        """
+        that ephemeral random overrides ceph.dir.bal.mask.
+        """
+        self.config_set('mds', 'mds_export_ephemeral_random', True)
+        self.config_set('mds', 'mds_export_ephemeral_random_max', 1.0)
+        count = 10
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "2")
+        for i in range(count):
+            self.mount_a.create_n_files("1/file", 100, sync=True)
+        self._wait_subtrees([('/1', 2)], status=self.status, rank=2)
+
+        self.mount_a.setfattr("1", "ceph.dir.pin.random", "1.0")
+        # timing issue
+        # Sometimes, there might be a delay in the synchronization of export_ephemeral_random_pin.
+        time.sleep(10)
+        for i in range(count):
+            self.mount_a.run_shell_payload(f"mkdir -p 1/{i}")
+            self.mount_a.create_n_files(f"1/{i}/file", 100, sync=True)
+
+        self._wait_random_subtrees(count, status=self.status, rank="all")
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "1")
+        for i in range(count):
+            self.mount_a.create_n_files("1/file", 100, sync=True)
+        self._wait_random_subtrees(count, status=self.status, rank="all")
+
+    def test_unset_rank_mask(self):
+        """
+        That ceph.dir.bal.mask is unset with -1.
+        """
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "2")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, "2")
+        self._wait_subtrees([('/1', 2)], status=self.status)
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "-1")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, None)
+
+    def test_unset_rank_mask_under_nested_root(self):
+        """
+        That ceph.dir.bal.mask is unset under root directory.
+        """
+
+        self.mount_a.setfattr(".", "ceph.dir.bal.mask", "0")
+        self._wait_subtrees([('', 0)], status=self.status, path='', exclude_path='~mds')
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "2")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, "2")
+        self._wait_subtrees([('/1', 2)], status=self.status)
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "-1")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, None)
+
+        self.mount_a.setfattr(".", "ceph.dir.bal.mask", "0")
+        self._wait_subtrees([('', 0)], status=self.status, path='', exclude_path='~mds')
```

Comment:
I'm not sure what this test is doing that `test_unset_rank_mask` isn't already testing?

Reply:
This unset test may be redundant, but '.' The directory is located at rank 0, and the '1' directory is exported at rank 2. The purpose is to test whether the parent directory's export affects the '1' directory.

Reply:
It seems the same line. https://github.com/ceph/ceph/pull/52373#discussion_r1532584706